ui/packages/security-servlet-logout/src/main/webapp/js/main.js

```diff
@@ -17,10 +14,68 @@
 
   $.get('../services/logout/actions', function(data) {
     var actions = JSON.parse(data)
-    processActions({ actions: actions })
-  })
-
-  $('#landinglink').click(function() {
-    window.location.href = window.location.href.replace(/logout\/.*/, '')
+    logout(actions).catch(error => {
+      console.log(error.message)
+      $('#error-msg').show()
+    })
   })
 })()
+
+async function logout(actions) {
+  let localLogoutAction
+  let externalLogoutActions = false
+  for (const action of actions) {
+    if (action.title === 'Local Logout') {
+      localLogoutAction = action
+    } else {
+      externalLogoutActions = true
+      // Certain protocols, e.g. SAML with the POST binding, use redirects that can't be followed
+      // with an AJAX request. For logout to work, we need to load the url in an iframe.
+      const iframe = document.createElement('iframe')
+      iframe.style.display = 'none'
+      iframe.src = action.url
+      document.body.appendChild(iframe)
+    }
+  }
+
+  // Wait 3 seconds for iframe logout
+  if (externalLogoutActions) {
+    await new Promise(r => setTimeout(r, 3000))
+  }
+
+  if (localLogoutAction) {
+    const response = await $.ajax(localLogoutAction.url)
+    if (response.mustCloseBrowser === true) {
+      $('#close-browser-msg').show()
+      return
+    }
+  }
+
+  const queryParams = parseQueryParams()
+  if (queryParams && queryParams['service']) {
+    $('#link').attr('href', queryParams['service'])
+  } else {
+    $('#link').attr('href', window.location.href.replace(/logout\/.*/, ''))
```

Comment:
❓ Doesn't this actually modify the `window.location.href` property? Are we wanting to do that?

Reply:
`window.location.replace` would, but this is just calling the String replace() method